### PR TITLE
Add documentation on batch

### PR DESCRIPTION
**Pull request by marikaner**
Improve and extend batch documentation.

### CODE REVIEW

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 1]
```diff
@@ -0,0 +1,60 @@
+
+A changeset is a collection of HTTP `POST`, `PUT`, `PATCH` and `DELETE` operations - requests built by any [CreateRequestBuilder](#create-request-builder), [UpdateRequestBuilder](#update-request-builder) and [DeleteRequestBuilder](#delete-request-builder) in terms of the SAP Cloud SDK.
+The order of execution within a changeset is not defined.
+This differs from the whole batch request itself, where the order is defined.
+Therefore the requests within a changeset should not depend on each other.
+If the execution of any of the requests within a changeset fails, the whole changeset will be reflected as an error in the response and will not be applied, much like a database transaction.
+
+Change request cannot be passed to a batch request directly.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:8]: Either 
> change requests cannot ...
or
> A change request cannot ...

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+They have to be combined in a changset, which in turn an be passed to the batch request.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:9]:
```suggestion
They have to be combined in a changset, which in turn can be passed to the batch request.
```

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+Once a batch request is executed it returns a list of `BatchResponse`s.
+Those contain the raw response information of each subrequest, the subresponse of a changeset request can either be `WriteResponses` or an `ErrorResponse`.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:11]:
```suggestion
Those contain the raw response information of each subrequest.
The subresponse of a changeset request can either be a `WriteResponse` or an `ErrorResponse`.
```

Comment (florian-richter) [reply docs/js/features/odata/common/batch/changeset.mdx:11]: I changed WriteResponses to the singular form. Not sure if the plural was intended!

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+To determine if a request was successful use `.isSuccess()`.
+
+Successful requests can be cast to `WriteResponses` which contains all subresponses for the changeset request.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:14]: In the JS context, how do you cast something to a type? (in typescript it's `as NEW_TYPE` but you can always do that, it's just sometimes wrong).

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+Those reponses can be accessed by `.responses` and have the type `WriteResponse`.
+Each `WriteResponse` contains the HTTP code and can contain the raw body and the constructor of the entity that was parsed from the response, depending on whether there was a response.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:16]: > , depending on whether there was a response.

How can the batch succeed if there was no response at all for a part of a changeset? I guess what is meant is that there the response had no body.

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+Create requests typically do not return a response.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:17]: response or body (aka `204 No Content` or `201 Created` is still a response in my understanding)

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+In order to work with an instance of an entity given in a `WriteResponse`, you can use the `.as` method, that allows you to transform the raw data into an instance of the given constructor.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:18]: Consider adding that the raw body is a string for people that may not be as experienced.

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+Note that the response cannot exist, so you should only call this method if you know that there is data.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:19]:
```suggestion
Note that the response may not exist, so you should only call this method if you know that there is data.
```

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+Typically the HTTP code is a good indicator for this (`201 No Content` probably won't have content).
+If you are working with TypeScript you will have to tell the compiler, that the `.as!` method can be used here by adding a `!`.
+Also note, that retrieve responses can be [`ErrorResponse`s](#error-handling).
+Therefore, it is crutial to check responses for success, before casting them to `WriteResponses`.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:23]:
```suggestion
Therefore, it is crucial to check responses for success, before casting them to `WriteResponses`.
```

File: docs/js/features/odata/common/batch/configure-serialization.mdx [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+GET /sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartnerAddress?$format=json HTTP/1.1
+```
+
+However, some services might only understand URLs relative to the entity or even absolute URLs.
+
+To configure the serialization of the URLs within a batch request, you can set the subrequest path type with `withSubRequestPathType`.
+You can either set it to `'relativeToService'`, as is the default, `'relativeToEntity'`, which will yield URLs relative to the entity or `'absolute'`, which will produce absolute URLs.
+See below for examples:
+
+Serialize subreqeust path relative to entity:
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/configure-serialization.mdx:12]:
```suggestion
Serialize subrequest path relative to entity:
```

File: docs/js/features/odata/common/batch/configure-serialization.mdx [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+```ts
+// GET /A_BusinessPartnerAddress?$format=json HTTP/1.1
+batch(...requests).withSubRequestPathType('relativeToEntity');
+```
+
+Serialize subreqeust path as absolute URL:
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/configure-serialization.mdx:18]:
```suggestion
Serialize subrequest path as absolute URL:
```

File: docs/js/features/odata/common/batch/serialization.mdx [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+POST /sap/opu/odata/sap/API_BUSINESS_PARTNER/A_BusinessPartnerAddress HTTP/1.1
+Content-Type: application/json
+Accept: application/json
+
+{"BusinessPartner":"1","PostalCode":"10001","City":"New York","Country":"USA"}
+
+```
+The first lines are request headers for the multipart request, followed by a blank line.
+The next line contains the request method and URL, followed by the request headers, a blank line and the request payload.
+Every "atomic" request is serialized to a string of this kind, while GET and DELETE requests do not provide a payload.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/serialization.mdx:16]:
```suggestion
Every "atomic" request is serialized to a string of this kind, while `GET` and `DELETE` requests do not provide a payload.
```

**Review by florian-richter — CHANGES_REQUESTED** [review]

File: docs/js/features/odata/common/batch/changeset.mdx (hunk 1 above)
Comment (marikaner) [inline docs/js/features/odata/common/batch/changeset.mdx:8]:
```suggestion
Change requests cannot be passed to a batch request directly.
```

File: docs/js/features/odata/common/batch/changeset.mdx (hunk 3 above)
Comment (marikaner) [reply docs/js/features/odata/common/batch/changeset.mdx:11]: No it was correct. I know it's weird 🙁

File: docs/js/features/odata/common/batch/changeset.mdx (hunk 4 above)
Comment (marikaner) [reply docs/js/features/odata/common/batch/changeset.mdx:14]: That is what is meant here, I will rephrase it to 'should'

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+If the execution of any of the requests within a changeset fails, the whole changeset will be reflected as an error in the response and will not be applied, much like a database transaction.
+
+Change request cannot be passed to a batch request directly.
+They have to be combined in a changset, which in turn an be passed to the batch request.
+Once a batch request is executed it returns a list of `BatchResponse`s.
+Those contain the raw response information of each subrequest, the subresponse of a changeset request can either be `WriteResponses` or an `ErrorResponse`.
+To determine if a request was successful use `.isSuccess()`.
+
+Successful requests can be cast to `WriteResponses` which contains all subresponses for the changeset request.
+Those reponses can be accessed by `.responses` and have the type `WriteResponse`.
```
Comment (marikaner) [inline docs/js/features/odata/common/batch/changeset.mdx:15]:
```suggestion
Those responses can be accessed by `.responses` and have the type `WriteResponse`.
```

File: docs/js/features/odata/common/batch/changeset.mdx (hunk 5 above)
Comment (marikaner) [inline docs/js/features/odata/common/batch/changeset.mdx:16]:
```suggestion
Each `WriteResponse` contains the HTTP code and can contain the raw body and the constructor of the entity that was parsed from the response, depending on whether there was a body in the response.
```

Comment (marikaner) [reply docs/js/features/odata/common/batch/changeset.mdx:16]: Correct, I adjusted it.

File: docs/js/features/odata/common/batch/changeset.mdx (hunk 6 above)
Comment (marikaner) [inline docs/js/features/odata/common/batch/changeset.mdx:17]:
```suggestion
Create and delete requests typically do not have a response body.
```

Comment (marikaner) [reply docs/js/features/odata/common/batch/changeset.mdx:17]: Yeah, I fixed it.

File: docs/js/features/odata/common/batch/changeset.mdx (hunk 7 above)
Comment (marikaner) [inline docs/js/features/odata/common/batch/changeset.mdx:18]:
```suggestion
In order to work with an instance of an entity given in a `WriteResponse`, you can use the `.as` method, that allows you to transform the raw string body into an instance of the given constructor.
```

**Comment by marikaner** [comment]
@florian-richter thanks, I fixed everything, feel free to check again.

File: docs/js/features/odata/common/batch/changeset.mdx [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+The order of execution within a changeset is not defined.
+This differs from the whole batch request itself, where the order is defined.
+Therefore the requests within a changeset should not depend on each other.
+If the execution of any of the requests within a changeset fails, the whole changeset will be reflected as an error in the response and will not be applied, much like a database transaction.
+
+Change requests cannot be passed to a batch request directly.
+They have to be combined in a changset, which in turn can be passed to the batch request.
+Once a batch request is executed it returns a list of `BatchResponse`s.
+Those contain the raw response information of each subrequest.
+The subresponse of a changeset request can either be `WriteResponses` or an `ErrorResponse`.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:12]: Maybe this can be improved by pointing out that there could be a collection of responses called `WriteResponses` or in case of failure a single `ErrorResponse`?

File: docs/js/features/odata/common/batch/changeset.mdx (hunk 3 above)
Comment (florian-richter) [reply docs/js/features/odata/common/batch/changeset.mdx]: Maybe this can be improved by pointing out that there could be a collection of responses called `WriteResponses` or in case of failure a single `ErrorResponse`?

File: docs/js/features/odata/common/batch/combining-requests.mdx [hunk 15]
```diff
@@ -0,0 +1,30 @@
+Batch requests can consist of as many retrieve requests and changesets as needed.
+In the example below, we combine create requests, update requests and get requests into one batch request.
+Remember, that any change requests have to be wrapped in a changeset.
+We put the create requests into one changeset and combine the update and delete requests into another.
+The get requests are added to the batch request directly without wrapping.
```
Comment (florian-richter) [inline docs/js/features/odata/common/batch/combining-requests.mdx:5]:
```suggestion
The `GET` requests are added to the batch request directly without wrapping.
```

**Review by florian-richter — APPROVED** [review]
See my comments regarding the "WriteResponses" and one improvement. Otherwise LGTM

File: docs/js/features/odata/common/batch/changeset.mdx (hunk 14 above)
Comment (florian-richter) [inline docs/js/features/odata/common/batch/changeset.mdx:12]: Maybe this can be improved by pointing out that there could be a collection of responses called `WriteResponses` or in case of failure a single `ErrorResponse`? (sorry if it was posted multiple times... github trolled me a bit)

Comment (marikaner) [inline docs/js/features/odata/common/batch/changeset.mdx:12]:
```suggestion
The response to a changeset request can either be a collection of the subresponses to the subrequests of the changeset of type `WriteResponses` or an `ErrorResponse`.
```

Comment (marikaner) [reply docs/js/features/odata/common/batch/changeset.mdx]: I rephrased it a bit. I found it difficult to express because the name of the type is not so great... Maybe something for the 2.0 discussion...
